src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -650,7 +653,13 @@ fn mtime_if_fresh<I>(output: &Path, paths: I) -> Option<FileTime>
     }
 }
 
-fn filename(unit: &Unit) -> String {
+fn filename(cx: &Context, unit: &Unit) -> String {
+    // If there exists a link stem, we have to use that since multiple target filenames
+    // may hardlink to the same target stem. If there's no link, we can use the original
+    // file_stem (which can include a suffix)
+    let file_stem = cx.link_stem(unit)
+        .map(|(_path, stem)| stem)
+        .unwrap_or_else(|| cx.file_stem(unit));
```

Comment:
Whoa! This is quite different than before (no nested fingerprint directories).

If feel like this is where all the magic, and I guess that's because `file_stem` includes a metadata hash in it?